File: src/piggytrade/trade_mapper.py
# trade_mapper.py
from dataclasses import dataclass
from typing import Optional

ERG = "ERG"

@dataclass
class TradeRoute:
    token_key: str      
    order_type: str     
    pool_type: str      

class TradeMapper:
    def __init__(self, tokens: dict):
        self.tokens = tokens
        self._erg_tokens = [k for k in tokens if not self._is_pair_key(k)]
        self._pair_tokens = [k for k in tokens if self._is_pair_key(k)]
        self._pair_sides: dict[str, tuple[str, str]] = {}  
        for key in self._pair_tokens:
            parts = key.split("-", 1)
            if len(parts) == 2:
                out_asset, in_asset = parts[0], parts[1]
                self._pair_sides[key] = (out_asset, in_asset)

    def all_assets(self) -> list:
        seen = set()
        seen.add(ERG)
        for k in self._erg_tokens:
            seen.add(k)
        for out_asset, in_asset in self._pair_sides.values():
            seen.add(out_asset)
            seen.add(in_asset)
        others = sorted(a for a in seen if a != ERG)
        return [ERG] + others

    def normalize_asset(self, asset_name: str) -> str:
        if not asset_name:
            return asset_name
        upper = asset_name.upper()
        if upper == "ERG":
            return ERG
        for asset in self.all_assets():
            if asset.upper() == upper:
                return asset
        return asset_name  

    def to_assets_for(self, from_asset: str) -> list:
        reachable = set()
        fa = self.normalize_asset(from_asset)
        if fa == ERG:
            for k in self._erg_tokens:
                reachable.add(k)
        elif fa in self._erg_tokens:
            reachable.add(ERG)
        for key, (out_asset, in_asset) in self._pair_sides.items():
            if fa == in_asset:
                reachable.add(out_asset)
            elif fa == out_asset:
                reachable.add(in_asset)
        if fa in self._erg_tokens:
            reachable.add(ERG)
        reachable.discard(fa)
        return sorted(reachable)

    def resolve(self, from_asset: str, to_asset: str) -> Optional[TradeRoute]:
        fa = self.normalize_asset(from_asset)
        ta = self.normalize_asset(to_asset)
        if fa == ERG and ta in self._erg_tokens:
            return TradeRoute(token_key=ta, order_type="BUY", pool_type="erg")
        if ta == ERG and fa in self._erg_tokens:
            return TradeRoute(token_key=fa, order_type="SELL", pool_type="erg")
        for key, (out_asset, in_asset) in self._pair_sides.items():
            if fa == in_asset and ta == out_asset:
                return TradeRoute(token_key=key, order_type="BUY", pool_type="token")
            if fa == out_asset and ta == in_asset:
                return TradeRoute(token_key=key, order_type="SELL", pool_type="token")
        return None  

    def describe_route(self, route: TradeRoute, amount: str, expected: str) -> str:
        # Pattern: {ReceivedAmount} {ReceivedAsset} for {SentAmount} {SentAsset}
        if route.pool_type == "erg":
            if route.order_type == "BUY":
                return f"{expected} {route.token_key} for {amount} ERG"
            else:
                return f"{expected} ERG for {amount} {route.token_key}"
        else:
            out_asset, in_asset = self._pair_sides.get(route.token_key, (route.token_key, "???"))
            if route.order_type == "BUY":
                return f"{expected} {out_asset} for {amount} {in_asset}"
            else:
                return f"{expected} {in_asset} for {amount} {out_asset}"

    @staticmethod
    def _is_pair_key(key: str) -> bool:
        return "-" in key

File: PiggyTradeNative/app/src/main/java/com/piggytrade/piggytrade/blockchain/TradeMapper.kt
package com.piggytrade.piggytrade.blockchain

import android.util.Log
import com.piggytrade.piggytrade.BuildConfig

data class TradeRoute(
    val tokenKey: String, // Original key from tokens map
    val orderType: String,
    val poolType: String,
    val pid: String = ""
)

class TradeMapper(private val tokens: Map<String, Map<String, Any>>) {

    companion object {
        const val ERG = "ERG"
        private const val TAG = "TradeMapper"
    }

    // Normalized asset names reachable directly from ERG
    private val ergTokens = mutableSetOf<String>()
    
    // Maps original token key -> Pair(n1, n2)
    private val poolAssetPairs = mutableMapOf<String, Pair<String, String>>()

    init {
        for ((key, data) in tokens) {
            val isT2T = data.containsKey("id_in") || data.containsKey("id_out")
            
            if (isT2T) {
                val nameIn = (data["name_in"] as? String)
                val nameOut = (data["name_out"] as? String)
                
                val parts = key.split("-", limit = 2)
                val rawN1 = nameIn ?: parts[0]
                val rawN2 = nameOut ?: if (parts.size > 1) parts[1] else ""
                
                val n1 = com.piggytrade.piggytrade.data.TokenRepository.normalizeTokenName(rawN1)
                val n2 = com.piggytrade.piggytrade.data.TokenRepository.normalizeTokenName(rawN2)
                
                poolAssetPairs[key] = Pair(n1, n2)
            } else {
                // ERG pool. The key is the identifier/name.
                ergTokens.add(com.piggytrade.piggytrade.data.TokenRepository.normalizeTokenName(key))
            }
        }
    }

    fun allAssets(): List<String> {
        val seen = mutableSetOf<String>()
        seen.add(ERG)
        seen.addAll(ergTokens)
        for (sides in poolAssetPairs.values) {
            seen.add(sides.first)
            seen.add(sides.second)
        }
        val others = seen.filter { it != ERG }.sorted()
        return listOf(ERG) + others
    }

    fun normalizeAsset(assetName: String?): String? {
        if (assetName.isNullOrBlank()) return assetName
        return com.piggytrade.piggytrade.data.TokenRepository.normalizeTokenName(assetName)
    }

    fun toAssetsFor(fromAsset: String): List<String> {
        val reachable = mutableSetOf<String>()
        val fa = normalizeAsset(fromAsset) ?: return emptyList()

        if (fa == ERG) {
            reachable.addAll(ergTokens)
        } else if (fa in ergTokens) {
            reachable.add(ERG)
        }

        for (sides in poolAssetPairs.values) {
            if (fa == sides.second) {
                reachable.add(sides.first)
            } else if (fa == sides.first) {
                reachable.add(sides.second)
            }
        }

        reachable.remove(fa)
        return reachable.sorted()
    }

    /**
     * Resolve exactly which POOL to use. 
     * If there are multiple pools for the same pair, this currently picks the FIRST one.
     * In a robust implementation, this would pick the most liquid one.
     */
    fun resolve(fromAsset: String, toAsset: String): TradeRoute? {
        val fa = normalizeAsset(fromAsset) ?: return null
        val ta = normalizeAsset(toAsset) ?: return null

        if (BuildConfig.DEBUG) Log.d(TAG, "resolve('$fromAsset','$toAsset') -> fa='$fa' ta='$ta' ergTokens=${ergTokens.size}")

        // 1. Check ERG-Token pools
        if (fa == ERG) {
            // Find a pool for ta where fa is ERG
            for ((key, data) in tokens) {
                if (data.containsKey("id_in")) continue // skip T2T
                val nToken = com.piggytrade.piggytrade.data.TokenRepository.normalizeTokenName(key)
                if (nToken == ta) {
                    val pid = data["pid"] as? String ?: ""
                    return TradeRoute(tokenKey = key, orderType = "BUY", poolType = "erg", pid = pid)
                }
            }
        }
        if (ta == ERG) {
            // Find a pool for fa where ta is ERG
            for ((key, data) in tokens) {
                if (data.containsKey("id_in")) continue // skip T2T
                val nToken = com.piggytrade.piggytrade.data.TokenRepository.normalizeTokenName(key)
                if (nToken == fa) {
                    val pid = data["pid"] as? String ?: ""
                    return TradeRoute(tokenKey = key, orderType = "SELL", poolType = "erg", pid = pid)
                }
            }
        }

        // 2. Check Token-Token pools
        for ((key, sides) in poolAssetPairs) {
            val pid = tokens[key]?.get("pid") as? String ?: ""
            if (fa == sides.second && ta == sides.first) {
                return TradeRoute(tokenKey = key, orderType = "BUY", poolType = "token", pid = pid)
            }
            if (fa == sides.first && ta == sides.second) {
                return TradeRoute(tokenKey = key, orderType = "SELL", poolType = "token", pid = pid)
            }
        }

        Log.w(TAG, "No route for '$fa' -> '$ta'.")
        return null
    }

    fun describeRoute(route: TradeRoute, amount: String, expected: String): String {
        val data = tokens[route.tokenKey]
        if (route.poolType == "erg") {
            return if (route.orderType == "BUY") {
                "$expected ${route.tokenKey} for $amount ERG"
            } else {
                "$expected ERG for $amount ${route.tokenKey}"
            }
        } else {
            val n1 = com.piggytrade.piggytrade.data.TokenRepository.normalizeTokenName(data?.get("name_in") as? String ?: route.tokenKey.split("-")[0])
            val n2 = com.piggytrade.piggytrade.data.TokenRepository.normalizeTokenName(data?.get("name_out") as? String ?: route.tokenKey.split("-").getOrNull(1) ?: "")
            
            return if (route.orderType == "BUY") {
                "$expected $n1 for $amount $n2"
            } else {
                "$expected $n2 for $amount $n1"
            }
        }
    }
}
